ntcore/ds_client.py

```diff
@@ -1,21 +1,112 @@
 
 import logging
 logger = logging.getLogger('nt')
 
+
 class DsClient(object):
     
     def __init__(self, dispatcher, verbose=False):
-        pass
-    
+        self.m_dispatcher = dispatcher
+        self.verbose = verbose
+
+        self.m_active = False
+        self.m_owner = None  # type: SafeThread
+
+        self.m_mutex = threading.Lock()
+        self.m_cond = Condition(self.m_mutex)
+
+        self.m_port = None  # type: int
+        self.m_stream = None
+
     def start(self, port):
-        logging.warning("Dashboard client mode not currently implemented")
-    
+        with self.m_mutex:
+            self.m_port = port
+            if not self.m_active:
+                self.m_active = True
+                self.m_owner = SafeThread(target=self._thread_main, name='nt-dsclient')
+
     def stop(self):
-        pass
+        with self.m_mutex:
+            # Close the stream so the read (if any) terminates.
+            self.m_active = False
+            if self.m_stream:
+                self.m_stream.close()
+            self.m_cond.notify()
+
+    def _thread_main(self):
+        oldip = 0
+        connector = TcpConnector(verbose=False, timeout=1)
+
+        while self.m_active:
+            # wait for periodic reconnect or termination
+            with self.m_mutex:
+                self.m_cond.wait_for(lambda: not self.m_active, timeout=0.5)
+                port = self.m_port
+
+            if not self.m_active:
+                break
+
+            self.m_stream = connector.connect(('127.0.0.1', 1742))
+            if not self.m_active:
+                break
+            if not self.m_stream:
+                continue
+
+            while self.m_active and self.m_stream:
+                json_blob = self.m_stream.readline()
+                if not json_blob:
+                    # We've reached EOF.
+                    with self.m_mutex:
+                        self.m_stream.close()
+                        self.m_stream = None
+
+                if not self.m_active:
+                    break
+
+                try:
+                    obj = json.loads(json_blob.decode())
+                except (json.JSONDecodeError, UnicodeDecodeError):
+                    continue
+                try:
+                    ip = obj['robotIP']
+                except KeyError:
+                    continue
+
+                # If zero, clear the server override
+                if ip == 0:
+                    self.m_dispatcher.clearServerOverride()
+                    oldip = 0
+                    continue
+
+                # If unchanged, don't reconnect
+                if ip == oldip:
+                    continue
+                oldip = ip
+
+                # Convert number into dotted quad
+                ip_str = '%d.%d.%d.%d' % ((ip >> 24) & 0xff, (ip >> 16) & 0xff,
+                                          (ip >> 8) & 0xff, ip & 0xff)
+                if self.verbose:
```

Comment:
Does this happen very often? I don't think this should be at verbose level.

Reply:
I recall seeing at competition that the DS will spin this between 10.TE.AM.2 and 0 when it's trying to connect. Probably a bad idea to spam in that scenario.